File: kernel/sbi/sbi.c
#include "sbi.h"

void sbi_put_char(char c) {
  register unsigned long a0 asm("a0") = (unsigned long)c;
  register unsigned long a7 asm("a7") = 1;
  asm volatile (
      "ecall"
      : "+r"(a0)
      : "r"(a7)
      : "memory"
  );
}

File: src/kernel/sbi.zig
pub fn readCsr(comptime reg: []const u8) u64 {
    return asm volatile ("csrr %[result], " ++ reg
        : [result] "=r" (-> u64),
    );
}

pub fn writeCsr(comptime reg: []const u8, val: u64) void {
    asm volatile (
        "csrw " ++ reg ++ ", %[value]"
        :
        : [value] "r" (val)
    );
}

pub fn putChar(ch: u8) void {
    asm volatile(
        "ecall"
        :
        : [ch]     "{a0}" (ch),
          [id]     "{a6}" (0),
          [legacy] "{a7}" (1),
        : "memory", "a0", "a1"
    );
}
